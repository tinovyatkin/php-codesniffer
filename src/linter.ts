/**
 * Launches phpCs and returns results as JSON
 */

import * as path from 'path';
import { promisify } from 'util';
import { exec } from 'child_process';
import * as assert from 'assert';

const execAsync = promisify(exec);

const EXECUTABLE_VERSIONS: Map<string, string> = new Map();
const DEFAULT_EXECUTABLE_PATH = path.resolve(
  __dirname,
  '../node_modules/php_codesniffer_master/bin/phpcs',
);

interface Options {
  encoding: string;
  standard: string;
}

const DEFAULT_OPTIONS: Partial<Options> = {
  encoding: 'UTF-8',
  standard: 'PEAR',
};

/**
 * Launches phpCs and returns current version
 */
export async function version(
  executablePath: string = DEFAULT_EXECUTABLE_PATH,
): Promise<string> {
  if (EXECUTABLE_VERSIONS.has(executablePath))
    return EXECUTABLE_VERSIONS.get(executablePath) as string;
  const { stdout } = await execAsync(`${executablePath} --version`, {
    windowsHide: true,
    timeout: 5000,
  });
  const ver = /^PHP_CodeSniffer version (?<ver>\d+\.\d+\.\d+)/i.exec(
    stdout.trim(),
  )?.groups?.ver;
  if (!ver)
    throw new ReferenceError(
      `Unknown version or invalid executable of phpcs, returned: "${stdout}"`,
    );
  EXECUTABLE_VERSIONS.set(executablePath, ver);
  return ver;
}

interface LinterMessage {
  message: string;
  source: string;
  severity: number;
  fixable: boolean;
  type: 'ERROR' | 'WARNING';
  line: number;
  column: number;
}
interface LintTotals {
  errors: number;
  warnings: number;
  fixable: number;
}

export async function lint(
  filenames: string | string[],
  executablePath = DEFAULT_EXECUTABLE_PATH,
  options = DEFAULT_OPTIONS,
): Promise<{
  totals: LintTotals;
  files: Record<
    string,
    { errors: number; warnings: number; message: LinterMessage[] }
  >;
}> {
  try {
    const ver = await version(executablePath);
    assert.ok(
      ver >= '2.6',
      `This library requires phpcs version 2.6 or later, received ${ver}`,
    );
    // we use promisified version, so, should not set exit code or it will throw
    const args = [
      '--report=json',
      '-q',
      `--encoding=${options.encoding}`,
      `--standard=${options.standard}`,
      '--runtime-set ignore_errors_on_exit 1',
      '--runtime-set ignore_warnings_on_exit 1',
    ];
    const { stdout } = await execAsync(
      `${executablePath} ${args.join(' ')} ${
        Array.isArray(filenames) ? filenames.join(' ') : filenames
      }`,
      {
        windowsHide: true,
        timeout: 15000,
      },
    );
    return JSON.parse(stdout);
  } catch (err) {
    if ('stdout' in err) {
      // Determine whether we have an error in stdout.
      const error = /^ERROR:\s?(?<error>.*)/i.exec(err.stdout)?.groups?.error;
      if (error) throw new Error(error.trim());
    }
    throw err;
  }
}
