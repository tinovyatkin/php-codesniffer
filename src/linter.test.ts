import * as path from 'path';

import { version, lint } from './linter';
describe('linter tests', () => {
  it('returns version', async () => {
    expect(await version('php ./test/phpcs.phar')).toBe('3.5.5');
  });
  it('throws on non-version', async () => {
    await expect(version('git')).rejects.toHaveProperty(
      'message',
      expect.stringContaining('Unknown version or invalid executable of phpcs'),
    );
  });

  it('lints several files', async () => {
    expect(
      await lint(
        './test/fixtures/test1.php ./test/fixtures/test2.php',
        undefined,
        {
          standard: path.resolve(
            __dirname,
            '../test/preferBeautifierConfig/subFolder/phpcs.xml',
          ),
        },
      ),
    ).toMatchSnapshot();
  });

  it('should parse phpcs errors and rethrow', async () => {
    await expect(lint('byaka')).rejects.toHaveProperty(
      'message',
      'The file "byaka" does not exist.',
    );
  });
});
