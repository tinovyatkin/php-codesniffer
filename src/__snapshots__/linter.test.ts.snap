// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linter tests lints several files 2`] = `
Array [
  Object {
    "errors": 8,
    "messages": Array [
      Object {
        "column": 1,
        "fixable": true,
        "line": 5,
        "message": "Inline control structures are not allowed",
        "severity": 5,
        "source": "Generic.ControlStructures.InlineControlStructure.NotAllowed",
        "type": "ERROR",
      },
      Object {
        "column": 13,
        "fixable": true,
        "line": 13,
        "message": "Newline required after opening brace",
        "severity": 5,
        "source": "Squiz.ControlStructures.ControlSignature.NewlineAfterOpenBrace",
        "type": "ERROR",
      },
      Object {
        "column": 1,
        "fixable": true,
        "line": 14,
        "message": "Expected 1 space after closing brace; 0 found",
        "severity": 5,
        "source": "Squiz.ControlStructures.ControlSignature.SpaceAfterCloseBrace",
        "type": "ERROR",
      },
      Object {
        "column": 7,
        "fixable": true,
        "line": 14,
        "message": "Newline required after opening brace",
        "severity": 5,
        "source": "Squiz.ControlStructures.ControlSignature.NewlineAfterOpenBrace",
        "type": "ERROR",
      },
      Object {
        "column": 47,
        "fixable": true,
        "line": 14,
        "message": "Closing brace must be on a line by itself",
        "severity": 5,
        "source": "Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore",
        "type": "ERROR",
      },
      Object {
        "column": 1,
        "fixable": true,
        "line": 23,
        "message": "Line indented incorrectly; expected at least 4 spaces, found 0",
        "severity": 5,
        "source": "Generic.WhiteSpace.ScopeIndent.Incorrect",
        "type": "ERROR",
      },
      Object {
        "column": 1,
        "fixable": true,
        "line": 24,
        "message": "Expected 1 space after closing brace; 0 found",
        "severity": 5,
        "source": "Squiz.ControlStructures.ControlSignature.SpaceAfterCloseBrace",
        "type": "ERROR",
      },
      Object {
        "column": 1,
        "fixable": true,
        "line": 25,
        "message": "Line indented incorrectly; expected at least 4 spaces, found 0",
        "severity": 5,
        "source": "Generic.WhiteSpace.ScopeIndent.Incorrect",
        "type": "ERROR",
      },
    ],
    "warnings": 0,
  },
  Object {
    "errors": 5,
    "messages": Array [
      Object {
        "column": 1,
        "fixable": false,
        "line": 1,
        "message": "A file should declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it should execute logic with side effects, but should not do both. The first symbol is defined on line 2 and the first side effect is on line 4.",
        "severity": 5,
        "source": "PSR1.Files.SideEffects.FoundWithSymbols",
        "type": "WARNING",
      },
      Object {
        "column": 16,
        "fixable": true,
        "line": 2,
        "message": "Expected 1 space between comma and argument \\"$b\\"; 5 found",
        "severity": 5,
        "source": "Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingBeforeArg",
        "type": "ERROR",
      },
      Object {
        "column": 26,
        "fixable": true,
        "line": 2,
        "message": "Opening brace should be on a new line",
        "severity": 5,
        "source": "Squiz.Functions.MultiLineFunctionDeclaration.BraceOnSameLine",
        "type": "ERROR",
      },
      Object {
        "column": 3,
        "fixable": true,
        "line": 3,
        "message": "Line indented incorrectly; expected at least 4 spaces, found 2",
        "severity": 5,
        "source": "Generic.WhiteSpace.ScopeIndent.Incorrect",
        "type": "ERROR",
      },
      Object {
        "column": 18,
        "fixable": true,
        "line": 3,
        "message": "Closing brace must be on a line by itself",
        "severity": 5,
        "source": "Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore",
        "type": "ERROR",
      },
      Object {
        "column": 13,
        "fixable": true,
        "line": 4,
        "message": "No space found after comma in argument list",
        "severity": 5,
        "source": "Generic.Functions.FunctionCallArgumentSpacing.NoSpaceAfterComma",
        "type": "ERROR",
      },
    ],
    "warnings": 1,
  },
]
`;
